sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt

```diff
@@ -513,23 +538,108 @@ class QProductCenterManager internal constructor(
-                it.onError(error)
+            retryLaunch(
+                onSuccess = {
+                    callbacks.handleSuccessProducts()
+                },
+                onError = { launchError ->
+                    val cachedLaunchResult = launchResultCache.getActualLaunchResult()
+
+                    if (cachedLaunchResult == null) {
+                        callbacks.handleFailureProducts(launchError)
+                    } else {
+                        callbacks.handleSuccessProducts(cachedLaunchResult.products)
+                    }
+                })
+        }
+    }
+
+    private fun List<QonversionProductsCallback>.handleSuccessProducts(
+        products: Map<String, QProduct>? = launchResult?.products
+    ) {
+        if (products != null) {
+            addSkuDetailForProducts(products.values)
+
+            this.forEach {
+                it.onSuccess(products)
             }
         }
     }
 
-    private fun productForID(id: String?): QProduct? {
-        return launchResult?.products?.get(id)
+    private fun List<QonversionProductsCallback>.handleFailureProducts(
+        error: QonversionError
+    ) {
+        this.forEach {
+            it.onError(error)
+        }
+    }
+
+    private fun handlePermissionsCallbacks(
+        callbacks: List<QonversionPermissionsCallback>
+    ) {
+        launchResult?.let {
+            val permissions = launchResult?.permissions ?: mapOf()
+            callbacks.handleSuccessPermissions(permissions)
+        } ?: run {
+            retryLaunch(
+                { launchResult ->
+                    callbacks.handleSuccessPermissions(launchResult.permissions)
+                },
+                { error ->
+                    if (forceLaunchRetry) {
+                        callbacks.handleFailurePermissions(error)
+                    } else {
+                        val cachedLaunchResult = launchResultCache.getActualLaunchResult()
+
+                        if (cachedLaunchResult == null) {
+                            callbacks.handleFailurePermissions(error)
+                        } else {
+                            callbacks.handleSuccessPermissions(cachedLaunchResult.permissions)
+                        }
+                    }
+                })
+        }
     }
 
+    private fun List<QonversionPermissionsCallback>.handleSuccessPermissions(
+        permissions: Map<String, QPermission>
+    ) {
+        this.forEach {
+            it.onSuccess(permissions)
+        }
+    }
+
+    private fun List<QonversionPermissionsCallback>.handleFailurePermissions(
+        error: QonversionError
+    ) {
+        this.forEach {
+            it.onError(error)
+        }
+    }
+
+    private fun productForID(id: String?): QProduct? = getAvailableLaunchResult()?.products?.get(id)
+
+
+    private fun getAvailableLaunchResult(): QLaunchResult? =
+        this@QProductCenterManager.launchResult ?: launchResultCache.getLaunchResult()
+
```

Comment:
The method is used for loadStoreProductsIfPossible() and productForID(). I use getLaunchResult() regardless of cache timestamp. Or we should use getActualLaunchResult() for the first version